### ISSUE TITLE
LVGL File Paths Change Could Trigger Dependant Projects to Update

### ISSUE THREAD

**Issue by vwheeler63** [kisvegabor/lvgl#7412]
### Introduce the problem

I have noticed that when files are added and/or removed from the LVGL repository that the `lv_port_pc_visual_studio` repository lags behind, sometimes for quite a while.  And until it gets updated, the project will not compile without modification.

One of the projects contained in that repository is `LVGL.MaintainerTools.sln` (solution file) => `LvglProjectFileUpdater.vcxproj` (project file), which, when run, uses the updated LVGL sub-module's working directory tree to update the list of files in the project (unfortunately, Visual Studio doesn't do that automatically for C/C++ projects).  This then makes the main projects (such as `LvglWindowsSimulator.vcxproj`) compile again.

I am doing that manually now when the directory structure gets updated.

The trouble is:  currently, due to this delay, new LVGL users have an (unwanted) high probability of running into this repository and trying it out during a time when it is not compiling, and this can unnecessarily damage LVGL public relations and goodwill among potential new users.

### Proposal

So my thoughts were some kind of trigger to get the `lv_port_pc_visual_studio` repository updated to match is needed in order to reduce the delay between the directory-structure update and the dependent repository(ies) being updated.  It can be anything that works:

- email to maintainer
- perhaps a GitHub notification of some sort to maintainer, e.g. with a new issue for that repository, so it gets timely attention(?)
- something automated (preferably not after every commit, but only when there is a change in the working directory tree)

If there are other similarly-dependent projects out there (I know Eclipse updates its projects automatically), then they might need such a trigger as well.

I'm guessing the maintainer of that repository would like to say what his needs and wants are for such a notification:  what would work well for him?

### Thoughts on Design

Something that is able to build a string representation of the directory structure (easy with Python `os.walk()`) could generate a hash (or "digest") of that string which would remain the same from commit to commit in the `lvgl/lvgl` repository, and when a file is added or removed, the hash would change and could thus provide a decision-point to trigger whatever kind of notification is most helpful.  Other than the script, the only thing required to store would be a file to store of the generated hash when it changes.

cc:  @kisvegabor @MouriNaruto

**Comment by vwheeler63** [comment]
> Can't we make Visual Studio to glob all the C and cpp files from the `lvgl/src` `lvgl/demos` `lvgl/examples` folder?

Unfortunately, Visual Stud8io is kind of stupid in that respect.  I think it is from a mis-guided effort on Microsoft's part to make the "tree structure" of files in the project to be INDEPENDENT of the actual directory structure.  Sadly, it is listed/discussed as a "feature" in forums about this issue.  That's why MouriNaruto had to write that "LvglProjectFileUpdater" program -- to do the job that Visual Studio isn't doing -- when it is dealing with a project whose source file structure can change frequently, and the changes are not done from within the VS project.

**Comment by vwheeler63** [comment]
> > Can't we make Visual Studio to glob all the C and cpp files from the `lvgl/src` `lvgl/demos` `lvgl/examples` folder?
> 
> Unfortunately, Visual Stud8io is kind of stupid in that respect. I think it is from a mis-guided effort on Microsoft's part to make the "tree structure" of files in the project to be INDEPENDENT of the actual directory structure. Sadly, it is listed/discussed as a "feature" in forums about this issue. That's why MouriNaruto had to write that "LvglProjectFileUpdater" program -- to do the job that Visual Studio isn't doing -- when it is dealing with a project whose source file structure can change frequently, and the changes are not done from within the VS project.

The result is that if, for example, a file `./lvgl/src/libs/path/to/lib/test_file.c` content is moved into another `.c` file and the file is deleted, suddenly the `lvgl/lv_port_pc_visual_studio` won't compile and requires its project being "fixed" by:

- manually by the user if he/she knows enough to do it [not intended], or
- by running the `LvglProjectFileUpdater` program (which updates 3 VS projects in the `lv_port_pc_visual_studio` repository.

**Comment by MouriNaruto** [comment]
I think I will try to write an MSBuild task to automate these tasks when people open the project.

Kenji Mouri

**Comment by vwheeler63** [comment]
> I think I will try to write an MSBuild task to automate these tasks when people open the project.
> 
> Kenji Mouri

Excellent!  Keep in mind that when updating the `lv_port_pc_visual_studio` repository locally, the update of the `lvgl` sub-module is sometimes automated and when the directory structure has changed, of course, it causes the 3 projects that rely on it to fail to compile with either "can't find file ..." or "symbol not found" errors depending on whether files went away, were added, or just moved or renamed.  So of course I am wondering if your idea keeps the projects in the repository updated, or whether it only updates the projects on the local workstations.  If the latter, I would prefer not having the delay and file re-writes every time I open it, since I open it frequently when no changes would apply (so it would be re-generating identical files over and over).

**Comment by MouriNaruto** [comment]
Not stale

**Comment by vwheeler63** [comment]
still not stale

**Comment by MouriNaruto** [comment]
not stale

**Comment by vwheeler63** [comment]
still not stale

**Comment by vwheeler63** [comment]
Not stale

**Comment by MouriNaruto** [comment]
Not stale

**Comment by vwheeler63** [comment]
Hi, @MouriNaruto !  FYI, a doc-structure change is about to be submitted in the next 24-48 hours or so (might be delayed by waiting on pending PRs to be merged).  I will give you a heads-up here so you have good prediction as soon as I do.

I'm kind of thinking we should start by simply sending you an email whenever one of us sees or knows about:

- new files
- changed names
- changed paths
- old files going away.

But thinking about that also tells me that somewhere we could keep a record of file structure on every commit to `master` have a Python script check it for structure change and use THAT to send an email.  What do you think?

**Comment by MouriNaruto** [comment]
@vwheeler63 

I think maybe the more better way in the current stage is providing batch scripts for users to update submodules.

I don't think it's a good idea for synchronizing from main repository for every changed commit because I think users may need the some stability.

Also, I think I need to write some tools help people migrate the lv_conf.h to new version. (I have found LVGL loves change this file a lot. I think if we need to provide users a up-to-date implementation, we should solve that issue.) If LVGL already has a prototype implementation for that will be much better.

Kenji Mouri

**Comment by vwheeler63** [comment]
> I think maybe the more better way in the current stage is providing batch scripts for users to update submodules.
> 
> I don't think it's a good idea for synchronizing from main repository for every changed commit because I think users may need the some stability.

If I think of myself doing UI development in this project. I would for sure create my own branch and that by itself would give me stability.  Then when I was ready to get new content, I would re-base on top of _________ (latest, preferred version, etc.).


> Also, I think I need to write some tools help people migrate the lv_conf.h to new version.

Good point!  I can imagine a Python script that starts with an array of everything the project should turn on and it could parse the new versions of `lv_conf_template.h` and generate new versions of `lv_conf.h` therewith....  If I were doing it, I would model it from `lvgl/docs/config_builder.py` but enhance its logic to read what needs to get "turned on" (0 => 1) driven by an external config file, perhaps governed by what the LVGL examples need.  Just my thought based on what I have seen/learned thus far.

**Comment by vwheeler63** [comment]
Not stale.

**Comment by MouriNaruto** [comment]
I think it's not stale.

But also we can talk this issue in https://github.com/lvgl/lv_port_pc_visual_studio/issues/101.

Kenji Mouri

**Comment by vwheeler63** [comment]
Not stale.

**Comment by MouriNaruto** [comment]
Not stale.

**Comment by vwheeler63** [comment]
Not stale.

**Comment by vwheeler63** [comment]
Not stale.

**Comment by MouriNaruto** [comment]
Not stale, I'm working on that.

Kenji Mouri

**Comment by AndreCostaaa** [comment]
Hi @MouriNaruto 

The other week I discussed with @vwheeler63 about this. Would you consider using `CMake` to generate the Visual Studio solution? Seems like a very straight forward and easy solution for the problem you're trying to solve here

**Comment by MouriNaruto** [comment]
@AndreCostaaa 

I don't consider CMake because we need proper out-of-box Visual Studio build experience.

Kenji Mouri

**Comment by vwheeler63** [comment]
> I don't consider CMake because we need proper out-of-box Visual Studio build experience.

@AndreCostaaa , @MouriNaruto has a good point because so many end users will be comfortable with Visual Studio, but will have no knowledge of CMake and so be placed in a very bad situation if anything goes wrong that breaks the compile, whereas in the normal VS environment, they will be armed to fix it.

**Comment by kisvegabor** [comment]
Adding a CI in the visual studio repo that updates the project every day or week if needed?

**Comment by MouriNaruto** [comment]
I think it's OK to add an CI after I finishing the project configuration migrator.

Kenji Mouri

**Comment by vwheeler63** [comment]
Not stale
29-Aug-2025 still not stale.

**Comment by AndreCostaaa** [comment]
I believe this issue belongs to https://github.com/lvgl/lv_port_pc_visual_studio as other port repositories don't suffer from the same problem

I still suggest using CMake to generate a visual studio solution, maybe based on a CI workflow

Let's reopen this if something comes up that should be done by the main repository

**Comment by vwheeler63** [comment]
> I believe this issue belongs to https://github.com/lvgl/lv_port_pc_visual_studio as other port repositories don't suffer from the same problem
> 
> I still suggest using CMake to generate a visual studio solution, maybe based on a CI workflow
> 
> Let's reopen this if something comes up that should be done by the main repository

Except.....   To send a trigger to that repository, it requires a change HERE.  Or is that even possible?  All it needs is a trigger from HERE (each merge) to send over to that REPO to do a check -- which I would be happy to write -- which would re-build the VS project to include the correct (new/changed) file structure and file names.  Otherwise, it causes the compile to fail.

After 9 months and it's not handled, I'm not confident that that repository has adequate attention.

What I was originally envisioning as a solution here is:
- Merge happens here, triggers "sending a trigger" to THAT repository.
- Script runs in that repository to check to see if file structure or names have changed (both of which cause the VS compile to fail).
- If a change HAS occurred, then the VS project is re-built (by the same script) and then, ideally, a commit is made, which....

Then automatically makes it possible to check out or clone that repository and have it compile successfully.  Lack of this ability makes it less usable because it has to be "repaired" by the end user before it works.

@AndreCostaaa

**Comment by MouriNaruto** [comment]
There are several issues when migrating the old lv_conf.h to the new lv_conf.h because there are more uncontrollable factors. But I know that the end developer want to edit that file directly.

Kenji Mouri

**Comment by MouriNaruto** [comment]
If the newer LVGL can keep the compatibility of the older lv_conf.h, it will be much easier, lol. I think I should assume that at this time because the pure regex cannot be able to satisfied with migrating the user-edited lv_conf.h.

Kenji Mouri

**Comment by MouriNaruto** [comment]
@AndreCostaaa 

> I still suggest using CMake to generate a visual studio solution, maybe based on a CI workflow

It's not OK for Visual Studio project users because Visual Studio already has CMake project support. CMake's Visual Studio project generator is only for compiling scenarios like the nmake generator and ninja generator.

**Comment by AndreCostaaa** [comment]
> Except..... To send a trigger to that repository, it requires a change HERE. Or is that even possible? All it needs is a trigger from HERE (each merge) to send over to that REPO to do a check

That workflow already exists, see:

https://github.com/lvgl/lvgl/blob/master/.github/workflows/release_branch_updater.yml

> There are several issues when migrating the old lv_conf.h to the new lv_conf.h

Have you seen lv_conf.defaults? It's how we solve this issue for the linux and vscode ports

> It's not OK for Visual Studio project users because Visual Studio already has CMake project support.

Can't you keep the CMakeLists and automatically generate the solution via the CI? It's an idea. I don't love it either since you'd be committing generated files but it would work

Anyway Visual Studio must have a way to update that list programmatically, right?

**Comment by MouriNaruto** [comment]
@AndreCostaaa 

> Have you seen lv_conf.defaults? It's how we solve this issue for the linux and vscode ports

I know that several months ago, but it seems not be designed for the end LVGL developers. If the end LVGL developers want to update their LVGL versions in their local repositories, it will be the issue because the most end LVGL developers won't use even notice the lv_conf.defaults.

Or learn from ESP32 toolchains which can only set the LVGL compilation options in their sdkconfig.defaults. But I don't think LVGL community and end developers will accept that because lv_conf.h is the only portable way for defining LVGL compilation options.

> Can't you keep the CMakeLists and automatically generate the solution via the CI?

It's not OK for CMake scripts to do complex XML generations, Visual Studio IDE itself has some issues that is unable to use wildcards for Visual Studio MSBuild C++ projects, this is why Visual Studio MSBuild C++ projects have the separate filters file. With my previous experience of maintaining the LVGL Visual Studio repository, all related users asked me must support show LVGL source folder files with the real file and folder structures in Visual Studio solution view a.k.a. filters view. Also, the modifying file items (for example, rename and remove) in the filters view is mandatory requirements. This is why I need to write generators with complex XML operations.

I think it's promising to provide a way for end LVGL developers to simplify and help them update the LVGL Visual Studio repository manually. This is the way I'm working on. It's not enough for us to only provide a CI in the LVGL main repository. 

> Anyway Visual Studio must have a way to update that list programmatically, right?

It's the best way to do that if you don't want to see lots of ugly unmaintainable CMake scripts, and it's suitable for LVGL Visual Studio repository users because it's MSBuild friendly. MSBuild can call .NET assembly or execute the inline C# source code directly.

I have tried my best to eliminate the C#-based logic, and use the MSBuild logic instead.

> I don't love it either since you'd be committing generated files but it would work.

I think my goal is making these process can be easily executed by end LVGL developers and GitHub Actions workflows. It can makes me only execute these when every stable LVGL release come out.

For example, I plans to provide some batch files which actually execute complex MSBuild based portable generator logics in the root folder of LVGL Visual Studio repository:

- UpdateDependencies.bat for updating submodules which is used in LVGL Visual Studio repository a.k.a. LVGL and FreeType currently, then synchronizing the Visual Studio MSBuild C++ projects filters view.
- MigrateConfigurations.bat for updating lv_conf.h from the lv_conf_template.h.

Kenji Mouri

**Comment by AndreCostaaa** [comment]
What I've done recently is use lv_conf_defaults as the source of truth and generate lv_conf.h as a build artifact. Would that be a possibility?

**Comment by MouriNaruto** [comment]
@AndreCostaaa 

> What I've done recently is use lv_conf_defaults as the source of truth and generate lv_conf.h as a build artifact. Would that be a possibility?

It will be easier for me if you can give end LVGL developers a good build configuration experience with the simpler configuration file and lv_conf.h is generated by that, which makes that I can write a proper migrator for that. (Even I don't need to write a lv_conf.h generator because I can use MSBuild custom task to read the configuration file to give all macro definitions to the compiler.)

I think maybe the ini style configuration file but with comments will be good for end LVGL developers.

Kenji Mouri
